zebra-state/src/service/tests.rs

```diff
@@ -314,6 +314,37 @@ proptest! {
             prop_assert_eq!(*best_tip_height.borrow(), Some(expected_height));
         }
     }
+
+    /// Test that the value pool is updated accordingly.
+    ///
+    /// 1. Generate a finalized chain and some non-finalized blocks.
+    /// 2. Check that initially the value pool is empty.
+    /// 3. Commit the finalized blocks and check that the value pool is updated accordingly.
+    /// 4. TODO: Commit the non-finalized blocks and check that the value pool is also updated
+    ///    accordingly.
+    #[test]
+    fn value_pool_is_updated(
+        (network, finalized_blocks, _non_finalized_blocks)
+            in continuous_empty_blocks_from_test_vectors(),
+    ) {
+        zebra_test::init();
+
+        let (mut state_service, _) = StateService::new(Config::ephemeral(), network);
+
+        use zebra_chain::value_balance::ValueBalance;
+        prop_assert_eq!(state_service.disk.current_value_pool(), ValueBalance::zero());
+
+        let mut expected_value_pool = Ok(ValueBalance::zero());
+        for block in finalized_blocks {
+            let utxos = &block.new_outputs;
+            let block_value_pool = &block.block.chain_value_pool_change(utxos)?;
+            expected_value_pool += *block_value_pool;
+
+            state_service.queue_and_commit_finalized(block);
+        }
+
+        prop_assert_eq!(state_service.disk.current_value_pool(), expected_value_pool?.constrain()?);
```

Comment:
This looks like a really good test!
Can we also check that the disk value pool is not zero at the end?

This extra check will catch bugs like skipping all the transactions in `chain_value_pool_change`.

Reply:
This is not always the case. This test seems to always be testing only the following value balances:

```
[zebra-state/src/service/tests.rs:348] state_service.disk.current_value_pool() = ValueBalance {
    transparent: Amount<zebra_chain::amount::NonNegative>(
        0,
    ),
    sprout: Amount<zebra_chain::amount::NonNegative>(
        0,
    ),
    sapling: Amount<zebra_chain::amount::NonNegative>(
        0,
    ),
    orchard: Amount<zebra_chain::amount::NonNegative>(
        0,
    ),
}
[zebra-state/src/service/tests.rs:363] network = Mainnet
[zebra-state/src/service/tests.rs:348] state_service.disk.current_value_pool() = ValueBalance {
    transparent: Amount<zebra_chain::amount::NonNegative>(
        937500,
    ),
    sprout: Amount<zebra_chain::amount::NonNegative>(
        0,
    ),
    sapling: Amount<zebra_chain::amount::NonNegative>(
        0,
    ),
    orchard: Amount<zebra_chain::amount::NonNegative>(
        0,
    ),
}
```

I think there may be some problems with `continuous_empty_blocks_from_test_vectors()` or i am not understanding the function correctly. For example, from my tests locally it is always returning the `Mainnet` and never the `Testnet` blocks.

Reply:
You're right, if `finalized_blocks` is empty, then the finalized value balance should be zero.

Otherwise, each block should add the block subsidy to the transparent pool.
(There are only 10 blocks, and Zcash transparent coinbase outputs can't be spent for 100 blocks, so all the other pools should be zero.)

I'm not sure about `Mainnet` and `Testnet`. 
Maybe `DEFAULT_PARTIAL_CHAIN_PROPTEST_CASES = 2` means that we never try `Testnet`?
(This could happen if it just runs the proptest seed cases.)

Do you want to just leave it here, and open a ticket to write more tests?
I'm happy with whatever you want to do here.

Reply:
The is returning 2 strategies, one with just the genesis block as finalized block and the other with the first 5 blocks including the genesis(what we want).

Ill see if i can get this fixed now in this PR.

Reply:
> The is returning 2 strategies, one with just the genesis block as finalized block and the other with the first 5 blocks including the genesis(what we want).

Sorry, in a second look into the function i think it is ok to have only the genesis block as finalized. It is just that with only two cases by default the prop tests seems to always try the lower edge first.

I think we can leave it as it is in this PR.

Reply:
Ok, I opened https://github.com/ZcashFoundation/zebra/issues/2640 in sprint 21, so we remember to write those tests when we finish off the value balance work.

Thanks for your work here, I think we're done with the code and tests in this PR.